### DISCUSSION TITLE
Possible to wrap ssh?

### DISCUSSION THREAD

**Discussion by chandlerkent** [Tyrrrz/CliWrap#208] [Q&A]
I am attempting to write a wrapper around open-ssh (ssh) and using CliWrap makes this much less daunting than using Process directly. However, I'm running into a few issues and want to know if anybody has experience doing this and does it work with CliWrap?

First here's my naive code:

```csharp
using CliWrap;
using CliWrap.Builders;

ArgumentsBuilder sshArgsBuilder = new();

// Pass-through any arguments
foreach (var arg in args)
{
	sshArgsBuilder.Add(arg);
}

var ssh = Cli.Wrap("ssh")
	.WithArguments(sshArgsBuilder.Build())
	.WithStandardOutputPipe(PipeTarget.ToStream(Console.OpenStandardOutput()))
	.WithStandardErrorPipe(PipeTarget.ToStream(Console.OpenStandardError()))
	.WithStandardInputPipe(PipeSource.FromStream(Console.OpenStandardInput()))
	.WithValidation(CommandResultValidation.None);

var cmdResult = await ssh.ExecuteAsync();
Environment.Exit(cmdResult.ExitCode);
```

When I run this trying to connect to some host I get the error:

    Pseudo-terminal will not be allocated because stdin is not a terminal.

Which I've seen other discussion on here on trying to pass this through `cmd` or otherwise. But when I try that I get the same error. Here's the code:

```csharp
using CliWrap;
using CliWrap.Builders;

ArgumentsBuilder sshArgsBuilder = new();

// Pass-through any additional arguments
foreach (var arg in args)
{
	sshArgsBuilder.Add(arg);
}

var ssh = Cli.Wrap("cmd")
	.WithArguments(args =>
		args
		.Add("/c")
		.Add($"ssh {sshArgsBuilder.Build()}")
	)
	.WithStandardOutputPipe(PipeTarget.ToStream(Console.OpenStandardOutput()))
	.WithStandardErrorPipe(PipeTarget.ToStream(Console.OpenStandardError()))
	.WithStandardInputPipe(PipeSource.FromStream(Console.OpenStandardInput()))
	.WithValidation(CommandResultValidation.None);

var cmdResult = await ssh.ExecuteAsync();
Environment.Exit(cmdResult.ExitCode);
```

OK, so I can work around this by using an argument in `ssh` called -t (or -tt to really force it?) and it gets rid of that error, but then I get duplicated output and prompts from ssh open a little dialog instead of allowing me to type into the shell directly.

For context, I am porting a similar tool I have written in nodejs to C# for _reasons_ and the nodejs version works simply with something like this:

```js
spawn(cmd, args, {
    stdio: [process.stdin, process.stdout, process.stderr],
  });
```

I'm not saying I expect CliWrap to work the same as nodejs here, but since I know it is possible for this to be done I'm wondering if CliWrap can do what I am expecting?

Thanks for the great library and thanks for any help!

**Comment by Tyrrrz** [comment]
Hi.

Honestly, I'm not sure. Never tried to wrap `ssh` myself. I'd really expect wrapping in a command shell of some sort to work, though.

Even then, I don't really understand why a pseudo-terminal is even necessary to read/write data from/to streams, but perhaps it's required for some `ssh`-specific behavior?

Maybe NodeJS does some special magic when you pipe the streams to `process.*`? Does it also work well when you pipe to some arbitrary buffer?

Also, could [SSH.NET](https://github.com/sshnet/SSH.NET) be a reasonable alternative? As far as I know, it doesn't actually wrap `ssh` or anything like that, but works with the protocol directly.

**Reply by chandlerkent** [reply]
Hey thanks for taking the time to reply! I'll try updating the nodejs version to write to some arbitrary buffer and let you know.

I investigated other options (like SSH.NET) but unfortunately our use case is kind of specific and nothing but open-ssh is supporting it (specifically we are using ssh certificates and support for those in other libraries is not existent).

**Reply by Tyrrrz** [reply]
I see. Not really sure what's going on inside `ssh`. Especially the "duplicate output" part sounds really weird.
